crates/rethnet_eth/src/remote.rs

```diff
@@ -0,0 +1,791 @@
+        /// block number
+        U64,
+        /// include transactions
+        bool,
+    ),
+    #[serde(rename = "eth_getCode")]
+    GetCode(
+        Address,
+        /// block number
+        U64,
+    ),
+    #[serde(rename = "eth_getTransactionCount")]
+    GetTxCount(
+        Address,
+        /// block number
+        U64,
+    ),
+}
+
+struct SingleRequest {
+    json: String,
+    id: jsonrpc::Id,
+}
+
+struct Response {
+    text: String,
+    request_body: String,
+    request_id: jsonrpc::Id,
+}
+
+struct BatchResponse {
+    text: String,
+    request_body: String,
+    request_ids: Vec<jsonrpc::Id>,
+}
+
+#[derive(serde::Serialize)]
+#[serde(rename_all = "camelCase")]
+struct GetLogsInput {
+    from_block: U64,
+    to_block: U64,
+    address: Address,
+}
+
+#[derive(serde::Serialize)]
+struct Request<'a> {
+    version: crate::remote::jsonrpc::Version,
+    #[serde(flatten)]
+    method: &'a MethodInvocation,
+    id: jsonrpc::Id,
+}
+
+impl RpcClient {
+    fn verify_success<T>(response: Response) -> Result<T, RpcClientError>
+    where
+        T: for<'a> serde::Deserialize<'a>,
+    {
+        let response_text = response.text.clone();
+        let success: jsonrpc::Success<T> = serde_json::from_str(&response.text).map_err(|err| {
+            RpcClientError::InterpretationError {
+                msg: err.to_string(),
+                request_body: response.request_body,
+                expected_type: format!(
+                    "rethnet_eth::remote::jsonrpc::Success<{}>",
+                    std::any::type_name::<T>().to_string()
+                ),
+                response_text,
+            }
+        })?;
+
+        assert_eq!(success.id, response.request_id);
+
+        Ok(success.result)
+    }
+
+    /// returns response text
+    async fn send_request_body(&self, request_body: String) -> Result<String, RpcClientError> {
+        use RpcClientError::{ResponseError, SendError};
+        self.client
+            .post(self.url.to_string())
+            .body(request_body.to_string())
+            .send()
+            .await
+            .map_err(|err| SendError {
+                msg: err.to_string(),
+                request_body: request_body.to_string(),
+            })?
+            .text()
+            .await
+            .map_err(|err| ResponseError {
+                msg: err.to_string(),
+                request_body: request_body.to_string(),
+            })
+    }
+
+    async fn call<T>(&self, input: &MethodInvocation) -> Result<T, RpcClientError>
+    where
+        T: for<'a> serde::Deserialize<'a>,
+    {
+        let id = jsonrpc::Id::Num(self.next_id.fetch_add(1, Ordering::Relaxed));
+        let SingleRequest { json, id } = SingleRequest {
+            json: serde_json::json!(Request {
+                version: crate::remote::jsonrpc::Version::V2_0,
+                id: id.clone(),
+                method: input,
+            })
+            .to_string(),
+            id,
+        };
```

Comment:
Do you need to construct the `SingleRequest` to then deconstruct it?

```suggestion
        let json = json: serde_json::json!(Request {
            version: crate::remote::jsonrpc::Version::V2_0,
            id: id.clone(),
            method: input,
        })
        .to_string();
```

Reply:
Thanks! This code has been refactored so many times, I'm happy to see it get cleaner haha. Addressed in https://github.com/NomicFoundation/hardhat/pull/3383/commits/6824c065738cef69d9123bf38d6394a8a0e254f0